text/0000-afp.md

```diff
@@ -0,0 +1,354 @@
+
+A farmer's response to a SOW, including cost per unit-of-work for the given SOW. 
+
+```proto
+message Quote {
+  bytes nonce = 1;
+  int64 per_unit_cost = 2;
+  SOW sow = 3;
+  AraId farmer = 4;
+}
+```
+
+#### 4.1.10. Agreement
+
+An agreed upon statement of work and quote between a farmer and a requester. This may reference a specific smart contract, or some other verifiable source, which can validate the details of an agreement.
+
+```proto
+message Agreement {
+  bytes nonce = 1;
+  Quote quote = 2;
+  Signature requester_signature = 3;
+  Signature farmer_signature = 4;
+  bytes data = 5;
+}
+```
+
+#### 4.1.11. Reward
+
+A reward for a specific agreement, sent by a verifiable peer.
+
+```proto
+message Reward {
+  bytes nonce = 1;
+	Agreement agreement = 2;
+  int64 amount = 3;
+  Signature requester_signature = 4;
+  bytes data = 5;
+}
+```
+
+#### 4.1.12. Receipt
+
+A receipt for a reward, sent by a verifiable peer.
+
+```proto
+message Receipt {
+  bytes nonce = 1;
+  Reward reward = 2;
+  Signature farmer_signature = 3;
+}
+```
+
+### 4.2. Interfaces
+
+In order to provide flexibility, this section describes interfaces that must be implemented on a per-service basis.
+
+#### 4.2.1. Requester
+
+A requester must extend the Requester class to define the requester's specifications for validating quotes for a task, creating and validating agreements for a task, and for starting a task.
+
+```js
+/**
+ * Returns whether a quote is valid.
+ * @param {messages.Quote} quote
+ * @returns {boolean}
+ */
+async validateQuote(quote) {
+  throw new Error('Extended classes must implement validateQuote.')
+}
+
+/**
+ * Generate and return an agreement for a quote.
+ * @param {messages.Quote} quote
+ * @returns {messages.Agreement}
+ */
+async generateAgreement(quote) {
+  throw new Error('Extended classes must implement generateAgreement.')
+}
+
+/**
+ * Return whether an agreement is valid.
+ * @param {messages.Agreement} agreement
+ * @returns {boolean}
+ */
+async validateAgreement(agreement) {
+  throw new Error('Extended classes must implement validateAgreement.')
+}
+
+/**
+ * Called when an agreement has been marked as valid and a farmer
+ * is ready to start work.
+ * @param {messages.Agreement} agreement
+ * @param {PeerConnection} connection
+ */
+async onHireConfirmed(agreement, connection) {
+  throw new Error('Extended classes must implement onHireConfirmed')
+}
+
+/**
+ * On receipt of a reward receipt from a farmer.
+ * @param {messages.Receipt} receipt
+ * @param {PeerConnection} connection
+ */
+async onReceipt(receipt, connection) {
+  throw new Error('Extended classes must implement onReceipt')
+}
+```
+
+#### 4.2.2. Farmer
+
+A farmer must extend the Farmer class to define that farmer’s specifications for generating a quote for a task, validating a SOW for a task, and signing and validating an agreement for a task. If rewards are expected, then a farmer would also handle validating the reward and generating a receipt.
```

Comment:
re: "If rewards are expected"- if`validateReward` and `generateReceipt` are required to be implemented, how does this affect rewards being optional?

Reply:
Hmm, if there aren't rewards sent, then those methods won't be called, so they don't need to implement them, though to be safe, you could implement them and just not have them do anything. Thoughts on how to state that?

Reply:
`throw new Error('If rewards are expected, extended classes must implement validateReward.')`